pyquil/quilbase.py

```diff
@@ -1120,22 +1124,98 @@ class Declare(AbstractInstruction):
+        sharing = None
+        if shared_region is not None:
+            sharing = quil_rs.Sharing(shared_region, Declare._to_rs_offsets(offsets))
+        return super().__new__(cls, name, vector, sharing)
+
+    @classmethod
+    def _from_rs_declaration(cls, declaration: quil_rs.Declaration) -> "Declare":
+        return super().__new__(cls, declaration.name, declaration.size, declaration.sharing)
+
+    @staticmethod
+    def _memory_type_to_scalar_type(memory_type: str) -> quil_rs.ScalarType:
+        memory_type = memory_type.upper()
+        if memory_type == "BIT":
+            return quil_rs.ScalarType.Bit
+        if memory_type == "INTEGER":
+            return quil_rs.ScalarType.Integer
+        if memory_type == "REAL":
+            return quil_rs.ScalarType.Real
+        if memory_type == "OCTET":
+            return quil_rs.ScalarType.Octet
+        raise ValueError(f"{memory_type} is not a valid scalar type.")
 
+    @staticmethod
+    def _to_rs_offsets(offsets: Optional[Iterable[Tuple[int, str]]]):
         if offsets is None:
-            offsets = []
-        self.offsets = offsets
+            return []
+        return [
+            quil_rs.Offset(offset, Declare._memory_type_to_scalar_type(memory_type)) for offset, memory_type in offsets
+        ]
+
+    @property
+    def memory_type(self) -> str:
+        return str(super().size.data_type)
+
+    @memory_type.setter
+    def memory_type(self, memory_type: str):
+        vector = super().size
+        vector.data_type = Declare._memory_type_to_scalar_type(memory_type)
+        quil_rs.Declaration.size.__set__(self, vector)
+
+    @property
+    def memory_size(self) -> int:
+        return super().size.length
+
+    @memory_size.setter
+    def memory_size(self, memory_size: int):
+        vector = super().size
+        vector.length = memory_size
+        quil_rs.Declaration.size.__set__(self, vector)
+
+    @property
+    def shared_region(self) -> Optional[str]:
+        sharing = super().sharing
+        if sharing is None:
+            return None
+        return sharing.name
+
+    @shared_region.setter
+    def shared_region(self, shared_region: Optional[str]):
+        sharing = super().sharing
+        if sharing is None:
+            if shared_region is None:
+                return
+            sharing = quil_rs.Sharing(shared_region, [])
+        else:
+            sharing.name = shared_region
+        quil_rs.Declaration.sharing.__set__(self, sharing)
+
+    @property
+    def offsets(self) -> List[Tuple[int, str]]:
+        sharing = super().sharing
+        if sharing is None:
+            return []
+        return [(offset.offset, str(offset.data_type)) for offset in sharing.offsets]
+
+    @offsets.setter
+    def offsets(self, offsets: Optional[List[Tuple[int, str]]]):
+        sharing = super().sharing
+        if sharing is None:
+            raise ValueError("DECLARE without a shared region cannot use offsets")
```

Comment:
This break is technically avoidable but would require us to setup `Sharing` with an empty string for a name, or leave it as none and discard the users `offsets`. Both felt undesirable to just enforcing correct usage, but I could be convinced otherwise.